packages/e2e-test-utils/src/insert-block.js

```diff
@@ -2,17 +2,39 @@
  * Internal dependencies
  */
 import { searchForBlock } from './search-for-block';
+import { getAllBlocks } from './get-all-blocks';
 
 /**
  * Opens the inserter, searches for the given term, then selects the first
- * result that appears.
+ * result that appears. It then waits briefly for the block list to update.
  *
  * @param {string} searchTerm The text to search the inserter for.
  */
 export async function insertBlock( searchTerm ) {
+	const oldBlocks = getAllBlocks();
+
 	await searchForBlock( searchTerm );
 	const insertButton = (
 		await page.$x( `//button//span[contains(text(), '${ searchTerm }')]` )
 	 )[ 0 ];
 	await insertButton.click();
+
+	const waitForBlocksToChange = ( delay = 20 ) =>
+		new Promise( ( resolve, reject ) => {
+			let elapsedTime = 0;
+			const pendingBlockList = setInterval( () => {
+				const blocks = getAllBlocks();
+				// Reference will change when the selector updates.
+				if ( blocks !== oldBlocks ) {
+					clearInterval( pendingBlockList );
+					resolve();
+				}
+				elapsedTime += delay;
+				if ( elapsedTime > 600 ) {
+					clearInterval( pendingBlockList );
+					reject( `Block ${ searchTerm } was never inserted.` );
+				}
+			}, delay );
+		} );
```

Comment:
To be entirely honest, I don't like this, the fact that it relies on "interval" and not "waitForSomething"

That said I know this PR has been lingering for some time and that you did some awesome work on it that I don't want to block it more than that.

Reply:
Thanks @youknowriad. I don’t disagree with you. It’s just that there doesn’t appear to be a way to compute a selector to wait for that will work for every case. It wouldn’t be a problem if the search term was the exact block name, but the search term can be anything and the function is just expected to return the first result.

I personally don’t think this code will hide real issues because it will fail within 600ms if a change was not detected. So it would still fail if a problem was introduced that made a block take very long to insert. 

I’m very happy to continue working on follow-up issues as we find them :) (and I’ll keep thinking about how to improve this test helper)

Reply:
600ms is a lot, I expect that if I insert a block and it's not there instantaneously (from the user's perspective), there's probably an issue somewhere for me.

Reply:
Why is there a delay in the blocks changing?

Reply:
I couldn't track it down. A few tests began failing which were all related to finding block UI after inserting a block. I investigated a lot of things and couldn't figure out why it would be failing (It's not noticeable to my eye performing test steps manually, or even watching the e2e tests run.) Maybe it is related to the memoization key change for `getBlocks`?

Reply:
for anyone reading this in the future, this was fixed in https://github.com/WordPress/gutenberg/pull/22377 by using `useLayoutEffect` instead of `useEffect` for the `InnerBlock` settings & template sync hooks.